x-pack/plugins/ingest_pipelines/public/application/components/pipeline_processors_editor/serialize.ts

```diff
@@ -15,35 +15,46 @@ export interface SerializeResult {
   on_failure?: Processor[];
 }
 
-const convertProcessorInternalToProcessor = (processor: ProcessorInternal): Processor => {
-  const { options, onFailure, type } = processor;
+const convertProcessorInternalToProcessor = (
+  processor: ProcessorInternal,
+  addTag?: boolean
+): Processor => {
+  const { options, onFailure, type, id } = processor;
   const outProcessor = {
     [type]: {
       ...options,
     },
   };
 
   if (onFailure?.length) {
-    outProcessor[type].on_failure = convertProcessors(onFailure);
-  } else if (onFailure) {
-    outProcessor[type].on_failure = [];
+    outProcessor[type].on_failure = convertProcessors(onFailure, addTag);
+  }
+
+  // For simulation, we add the "tag" field equal to the internal processor id
+  // so that we can map the simulate results to each processor
+  if (addTag) {
+    outProcessor[type].tag = id;
   }
 
   return outProcessor;
 };
 
-const convertProcessors = (processors: ProcessorInternal[]) => {
+const convertProcessors = (processors: ProcessorInternal[], addTag?: boolean) => {
   const convertedProcessors = [];
 
   for (const processor of processors) {
-    convertedProcessors.push(convertProcessorInternalToProcessor(processor));
+    convertedProcessors.push(convertProcessorInternalToProcessor(processor, addTag));
   }
+
   return convertedProcessors;
 };
 
-export const serialize = ({ processors, onFailure }: SerializeArgs): SerializeResult => {
+export const serialize = (
```

Comment:
nit; a default value and renaming to `copyIdToTag` could be extra clear :)

Then we document like

```ts
/**
  * param@ processor: The deserialized processor to convert
  * param@ copyIdToTag: For simulation, we add the "tag" field equal to the internal processor id so that we can map the simulate results to each processor
  */
const serialize = (processor: ProcessorInternal, copyIdToTag: boolean = false)
```

I think doing this on interface could be even better, but understand that is more of a refactor:

```ts
interface SerializeArgs {
  /**
     * The deserialized processor to convert
     */
  processor: ProcessorInternal;

  /**
     * For simulation, we add the "tag" field equal to the internal processor id so that we can map the simulate results to each processor
     */
  copyIdToTag: boolean;
}

export const serialize = ({
  pipeline: { processors, onFailure },
  copyIdToTag = false,
}: SerializeArgs): SerializeResult => {

```